lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -1640,17 +1654,118 @@ namespace Js
+        int32_t length = 0;
+        if (langtag == nullptr)
         {
-            JavascriptError::MapAndThrowError(scriptContext, E_INVALIDARG);
+            length = uloc_getName(nullptr, localeID, _countof(localeID), &error);
         }
+        else
+        {
+            char langtag8[ULOC_FULLNAME_CAPACITY] = { 0 };
+            AssertOrFailFast(utf8::WideStringToNarrowNoAlloc(langtag, langtagLen, langtag8, _countof(langtag8)) == S_OK);
+            uloc_forLanguageTag(langtag8, localeID, _countof(localeID), &length, &error);
+        }
+        ICU_ASSERT(error, length > 0);
+
+        ScopedUCollator collator(ucol_open(localeID, &error));
+        ICU_ASSERT(error, true);
 
+        if (sensitivity == CollatorSensitivity::Base)
+        {
+            ucol_setStrength(collator, UCOL_PRIMARY);
+        }
+        else if (sensitivity == CollatorSensitivity::Accent)
+        {
+            ucol_setStrength(collator, UCOL_SECONDARY);
+        }
+        else if (sensitivity == CollatorSensitivity::Case)
+        {
+            // see "description" for the caseLevel default option: http://userguide.icu-project.org/collation/customization
+            ucol_setStrength(collator, UCOL_PRIMARY);
+            ucol_setAttribute(collator, UCOL_CASE_LEVEL, UCOL_ON, &error);
+            ICU_ASSERT(error, true);
+        }
+        else if (sensitivity == CollatorSensitivity::Variant)
+        {
+            ucol_setStrength(collator, UCOL_TERTIARY);
+        }
+        else
+        {
+            AssertMsg(false, "sensitivity is not one of the CollatorSensitivity values");
+        }
+
+        if (ignorePunctuation)
+        {
+            // see http://userguide.icu-project.org/collation/customization/ignorepunct
+            ucol_setAttribute(collator, UCOL_ALTERNATE_HANDLING, UCOL_SHIFTED, &error);
+            ICU_ASSERT(error, true);
+        }
+
+        if (numeric)
+        {
+            ucol_setAttribute(collator, UCOL_NUMERIC_COLLATION, UCOL_ON, &error);
+            ICU_ASSERT(error, true);
+        }
+
+        if (caseFirst == CollatorCaseFirst::Upper)
+        {
+            ucol_setAttribute(collator, UCOL_CASE_FIRST, UCOL_UPPER_FIRST, &error);
+            ICU_ASSERT(error, true);
+        }
+        else if (caseFirst == CollatorCaseFirst::Lower)
+        {
+            ucol_setAttribute(collator, UCOL_CASE_FIRST, UCOL_LOWER_FIRST, &error);
+            ICU_ASSERT(error, true);
+        }
+
+        *hr = S_OK;
+        UCollationResult result = ucol_strcoll(collator, reinterpret_cast<const UChar *>(left), cchLeft, reinterpret_cast<const UChar *>(right), cchRight);
+        if (result == UCOL_LESS)
+        {
+            ret = 1;
+        }
+        else if (result == UCOL_EQUAL)
+        {
+            ret = 2;
+        }
+        else if (result == UCOL_GREATER)
+        {
+            ret = 3;
+        }
+        else
+        {
+            *hr = E_FAIL;
+            ret = 0;
+        }
```

Comment:
out of curiosity what would cause this? Will the caller handle hr correctly? If not (future proofing) what will it do with ret == 0?

Reply:
Well, the function is used in place of CompareStringEx, which returns 0 on fail. So it goes through the same path where if ret == 0, we MapAndThrowError(hr)